contracts/utils/cryptography/ZKEmailUtils.sol

```diff
@@ -0,0 +1,145 @@
+ * mechanism to ensure the email command matches the expected format and parameters.
+ * * A https://docs.zk.email/architecture/zk-proofs#how-zk-email-uses-zero-knowledge-proofs[zero-knowledge proof] verification
+ * mechanism to ensure the email was actually sent and received without revealing its contents. Defined by an `IVerifier` interface.
+ */
+library ZKEmailUtils {
+    using Strings for string;
+
+    /// @dev Enumeration of possible email proof validation errors.
+    enum EmailProofError {
+        NoError,
+        DKIMPublicKeyHash, // The DKIM public key hash verification fails
+        MaskedCommandLength, // The masked command length exceeds the maximum
+        SkippedCommandPrefixSize, // The skipped command prefix size is invalid
+        MismatchedCommand, // The command does not match the proof command
+        EmailProof // The email proof verification fails
+    }
+
+    /// @dev Enumeration of possible string cases used to compare the command with the expected proven command.
+    enum Case {
+        LOWERCASE, // Converts the command to hex lowercase.
+        UPPERCASE, // Converts the command to hex uppercase.
+        CHECKSUM // Computes a checksum of the command.
+    }
+
+    /// @dev Variant of {isValidZKEmail} that validates the `["signHash", "{uint}"]` command template.
+    function isValidZKEmail(
+        EmailAuthMsg memory emailAuthMsg,
+        IDKIMRegistry dkimregistry,
+        IVerifier verifier
+    ) internal view returns (EmailProofError) {
+        string[] memory signHashTemplate = new string[](2);
+        signHashTemplate[0] = "signHash";
+        signHashTemplate[1] = CommandUtils.UINT_MATCHER;
+
+        // UINT_MATCHER is always lowercase
+        return isValidZKEmail(emailAuthMsg, dkimregistry, verifier, signHashTemplate, Case.LOWERCASE);
+    }
+
+    /**
+     * @dev Validates a ZKEmail authentication message.
+     *
+     * This function takes an email authentication message, a DKIM registry contract, and a verifier contract
+     * as inputs. It performs several validation checks and returns a tuple containing a boolean success flag
+     * and an {EmailProofError} if validation failed. See {validateDKIMAndCommandFormat} and {validateExpectedCommandAndProof}
+     * for more details on the validation checks performed.
+     *
+     * NOTE: Attempts to validate the command for all possible string {Case} values.
+     */
+    function isValidZKEmail(
+        EmailAuthMsg memory emailAuthMsg,
+        IDKIMRegistry dkimregistry,
+        IVerifier verifier,
+        string[] memory template
+    ) internal view returns (EmailProofError) {
+        EmailProofError err = validateDKIMAndCommandFormat(emailAuthMsg, dkimregistry, verifier);
+        if (err != EmailProofError.NoError) return err;
+        for (uint256 i = 0; i < uint8(type(Case).max) && err != EmailProofError.NoError; i++) {
```

Comment:
That change is still broken. If `err` is `NoError` you don't enter the loop.

Reply:
I'll try the alternative approach with `_commandMatch`. Next time please write a suggestion, it's difficult to follow "if this then that, but then this, otherwise we should do this other", I understand the function is broken.